Sources/Basics/Environment/Environment.swift

```diff
@@ -0,0 +1,326 @@
+        Self._cachedCurrent.withLock { cachedValue in
+            if let cachedValue = cachedValue {
+                return cachedValue
+            } else {
+                let current = Self(ProcessInfo.processInfo.environment)
+                cachedValue = current
+                return current
+            }
+        }
+    }
+
+    /// Temporary override environment variables
+    ///
+    /// WARNING! This method is not thread-safe. POSIX environments are shared
+    /// between threads. This means that when this method is called simultaneously
+    /// from different threads, the environment will neither be setup nor restored
+    /// correctly.
+    package static func makeCustom<T>(
+        _ environment: Self,
+        body: () async throws -> T
+    ) async throws -> T {
+        let current = Self.current
+        let state = environment.storage.keys.map { ($0, current[$0]) }
+        let restore = {
+            for (key, value) in state {
+                try Self.set(key: key, value: value)
+            }
+        }
+        let returnValue: T
+        do {
+            for (key, value) in environment {
+                try Self.set(key: key, value: value)
+            }
+            returnValue = try await body()
+        } catch {
+            try? restore()
+            throw error
+        }
+        try restore()
+        return returnValue
+    }
+
+    /// Temporary override environment variables
+    ///
+    /// WARNING! This method is not thread-safe. POSIX environments are shared
+    /// between threads. This means that when this method is called simultaneously
+    /// from different threads, the environment will neither be setup nor restored
+    /// correctly.
+    package static func makeCustom<T>(
+        _ environment: Self,
+        body: () throws -> T
+    ) throws -> T {
+        let current = Self.current
+        let state = environment.storage.keys.map { ($0, current[$0]) }
+        let restore = {
+            for (key, value) in state {
+                try Self.set(key: key, value: value)
+            }
+        }
+        let returnValue: T
+        do {
+            for (key, value) in environment {
+                try Self.set(key: key, value: value)
+            }
+            returnValue = try body()
+        } catch {
+            try? restore()
+            throw error
+        }
+        try restore()
+        return returnValue
+    }
+
+    struct UpdateEnvironmentError: CustomStringConvertible, Error {
+        var function: StaticString
+        var code: Int32
+        var description: String { "\(self.function) returned \(self.code)" }
+    }
+
+    /// Modifies the process's global environment.
+    ///
+    /// > Important: This operation is _not_ concurrency safe.
+    package static func set(key: EnvironmentKey, value: String?) throws {
+        if let value = value {
+            #if os(Windows)
+            guard _putenv("\(key)=\(value)") == 0 else {
```

Comment:
This is annoying - I don't know if we should use `_putenv` or `SetEnvironmentVariableW`. The former only adjusts the "C" functions view, the latter only changes the Win32 view. I suppose that we could call both?

Reply:
I can have it call both, in general I strongly dislike mutating the global environment anyways